deps/picotls/lib/openssl.c

```diff
@@ -693,77 +695,138 @@ int ptls_openssl_create_key_exchange(ptls_key_exchange_context_t **ctx, EVP_PKEY
             goto Exit;
         }
-    } else
-#endif
-    {
-        if (EVP_PKEY_id(key) == EVP_PKEY_RSA) {
-            if (EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, RSA_PKCS1_PSS_PADDING) != 1) {
+
+#if PTLS_OPENSSL_HAVE_ED25519
+        if (EVP_PKEY_id(key) == EVP_PKEY_ED25519) {
+            /* ED25519 requires the use of the all-at-once function that appeared in OpenSSL 1.1.1, hence different path */
+            if (EVP_DigestSign(args->ctx, NULL, &args->siglen, input.base, input.len) != 1) {
                 ret = PTLS_ERROR_LIBRARY;
                 goto Exit;
             }
-            if (EVP_PKEY_CTX_set_rsa_pss_saltlen(pkey_ctx, -1) != 1) {
+            if ((ret = ptls_buffer_reserve(outbuf, args->siglen)) != 0)
+                goto Exit;
+            if (EVP_DigestSign(args->ctx, outbuf->base + outbuf->off, &args->siglen, input.base, input.len) != 1) {
                 ret = PTLS_ERROR_LIBRARY;
                 goto Exit;
             }
-            if (EVP_PKEY_CTX_set_rsa_mgf1_md(pkey_ctx, md) != 1) {
+        } else
+#endif
+        {
+            if (EVP_PKEY_id(key) == EVP_PKEY_RSA) {
+                if (EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, RSA_PKCS1_PSS_PADDING) != 1) {
+                    ret = PTLS_ERROR_LIBRARY;
+                    goto Exit;
+                }
+                if (EVP_PKEY_CTX_set_rsa_pss_saltlen(pkey_ctx, -1) != 1) {
+                    ret = PTLS_ERROR_LIBRARY;
+                    goto Exit;
+                }
+                if (EVP_PKEY_CTX_set_rsa_mgf1_md(pkey_ctx, md) != 1) {
+                    ret = PTLS_ERROR_LIBRARY;
+                    goto Exit;
+                }
+            }
+            if (EVP_DigestSignUpdate(args->ctx, input.base, input.len) != 1) {
                 ret = PTLS_ERROR_LIBRARY;
                 goto Exit;
             }
-        }
-        if (EVP_DigestSignUpdate(ctx, input.base, input.len) != 1) {
-            ret = PTLS_ERROR_LIBRARY;
-            goto Exit;
-        }
-        if (EVP_DigestSignFinal(ctx, NULL, &siglen) != 1) {
-            ret = PTLS_ERROR_LIBRARY;
-            goto Exit;
-        }
-        if ((ret = ptls_buffer_reserve(outbuf, siglen)) != 0)
-            goto Exit;
-        if (EVP_DigestSignFinal(ctx, outbuf->base + outbuf->off, &siglen) != 1) {
-            ret = PTLS_ERROR_LIBRARY;
-            goto Exit;
+
+            if (EVP_DigestSignFinal(args->ctx, NULL, &args->siglen) != 1) {
+                ret = PTLS_ERROR_LIBRARY;
+                goto Exit;
+            }
+
+            if ((ret = ptls_buffer_reserve(&args->buf, args->siglen)) != 0)
+                goto Exit;
         }
     }
 
-    outbuf->off += siglen;
+#if PTLS_OPENSSL_HAVE_ED25519
+    if (EVP_PKEY_id(key) == EVP_PKEY_ED25519) {
+        // signing is complete as ED25519 is one-shot
+    } else
+#endif
+    {
+        if (ASYNC_get_current_job() == NULL) {
+            if (args->waitctx == NULL) {
+                args->waitctx = ASYNC_WAIT_CTX_new();
```

Comment:
Would it make sense to inject `ASYNC_WAIT_CTX` from the application rather than lazy-initializing it inside picotls?

I ask this because to me it seems that `ASYNC_WAIT_CTX` is designed so that it would be shared among multilple async jobs. When picotls is only one of the module using OpenSSL, it might be better for the application to specify a context that is shared among all the modules using OpenSSL.

Regarding how, I do not have a strong opinion. It could be a member of `ptls_openssl_sign_certificate_t`, or a callback of `ptls_openssl_sign_certificate_t`. Or else, we can export a function that is called from a user-defined `sign_certificate` callback. This exported function would take `ASYNC_WAIT_CTX *` as an argument in addition to all the necessary parameters for generating a signature.

Reply:
The documentation [here](https://www.openssl.org/docs/man3.0/man3/ASYNC_WAIT_CTX_new.html) leads me to a different interpretation, notably:

> An ASYNC_WAIT_CTX object represents an asynchronous "session", i.e. a related set of crypto operations. For example in SSL terms this would have a one-to-one correspondence with an SSL connection.

>When the job is started it is associated with the ASYNC_WAIT_CTX for the duration of that job. An ASYNC_WAIT_CTX should only be used for one ASYNC_JOB at any one time, but can be reused after an ASYNC_JOB has finished for a subsequent ASYNC_JOB. When the session is complete (e.g. the SSL connection is closed), application code cleans up with ASYNC_WAIT_CTX_free()

Here is how OpenSSL uses it internally (stored on `SSL *`): https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/ssl/ssl_lib.c#L1678-L1710

What if we kept the lazy initialisation but stored the `waitctx` and the `job` on `ptls_openssl_sign_certificate_t` instead?

Reply:
Thank you for the explanation. I concur.

I now think it is fine and makes sense to have `ptls_openssl_get_async_fd`.

Regarding how we instantiate `ASYNC_WAIT_CTX`, would it make sense to do that inside `sign_ctx_alloc` based on my renewed understanding that we'd be using one instance per signature. The same argument might apply to `EVP_MD_CTX`. The potential benefit is that then these pointers would essentially have the guarantee to be non-NULL across our code.

Reply:
`ASYNC_WAIT_CTX` yes, moved to `sign_ctx_alloc`

For `EVP_MD_CTX`: we need a `ctx` to get a `siglen` before we alloc(siglen), we then transfer ownership of `EVP_MD_CTX` to the `sign_ctx`

Reply:
addressed in https://github.com/h2o/picotls/pull/408